### ISSUE TITLE
How to know the table and column name that is the problem in a foreign key exception in Entity Core?

### ISSUE THREAD

**Issue by ComptonAlvaro** [dotnet/efcore#32476]
Hello:

I am using entity Core 8.0 and to see how to handle the exceptions of Entity Core when I am using SQL Server.

For that, I am creating a helper that has a GetException() method that gets an exception as a parameter and tries to analyze the exception to give a more specific exception, to provide more information to the client.

At first, I am trying to use this helper in the SaveChanges() method, to use it only in one place.

In the case of the foreign key exception, when I try to assign an FK that doesn’t exist, I get the error number 547, and the message is this: “The INSERT statement conflicted with the FOREIGN KEY constraint ‘NameOfRelationship’. The conflict occurred in database ‘MyDatabaseName’, table ‘dbo.Mytable’, column ‘IDForeignKey’.”

The problem is that the text is in Spanish, so parsing the message doesn’t seem to be a good option, because if the application runs on another computer with another culture, I would have problems.

It is true that the repository runs on the server, and I can control the language that is used on the server, but if there were another option instead of parsing the message, it would be good.

This problem is because I use the helper in the SaveChanges() method, so I am getting any foreign key of every table in the database.

One option would be to use this helper in each method of the repository, so I could search for the name of all the possible FKs in the tables that are affected by the method. But this would make me do the same in all the methods of the repository. That is why I am trying to use it only in the SaveChanges() method and try to find the column with the problem in some way.

But if the message error, EF Core tells the name of the relationship, the table name and the name of the property. is this information in some property or in some way that I could check to avoid the need to parse the message?

So, in summary, I would like to know if there is some way to get the information from a different way than from the message. Or perhaps this helper is not a good idea and there is another better way to handle exceptions in Entity Core.

Thanks.

**Comment by roji** [comment]
@ComptonAlvaro the exception text that you're seeing comes from the low-level database driver (e.g. SqlClient), not from EF Core; EF simply generates SQL for you, sends it to the databases (via SqlClient), and bubbles up any exceptions that SqlClient throws.

PostgreSQL specifically does raise the name of the unique constraint which was violated, and Npgsql exposes that information via [PostgresException.ConstraintName](https://www.npgsql.org/doc/api/Npgsql.PostgresException.html#Npgsql_PostgresException_ConstraintName). I don't think SQL Server/SqlClient expose such information in a similar way, but you may want to check with SqlClient to be sure.

I'm generally not sure that trying to write such a helper makes much sense; the database (and its driver) provide the information which they provide - I wouldn't try to parse/infer/add information that isn't really there (in a structured way).
